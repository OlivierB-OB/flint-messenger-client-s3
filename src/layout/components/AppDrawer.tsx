import Box from '@material-ui/core/Box';
import Drawer from '@material-ui/core/Drawer';
import IconButton from '@material-ui/core/IconButton';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import ArrowBackIos from '@material-ui/icons/ArrowBackIos';
import React from 'react';
import { connect } from 'react-redux';
import { Dispatch, Action } from 'redux';
import { IAppState } from '../../appReducer';
import { IDrawerContent } from '../types';
import { hideDrawer } from '../actions/hideDrawer';
import { MyContacts } from '../../users/components/MyContacts';
import { MyConversations } from '../../conversations/components/MyConversations';

export interface IDrawerDisplayProps {
  show: boolean;
  content?: IDrawerContent;
  hideDrawer: () => void;
}

export const drawerWidth = 500;

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    drawerHeader: {
      textAlign: 'right',
      position: 'sticky',
      top: 0,
      zIndex: 100,
      backgroundColor: theme.palette.background.paper,
      borderBottom: `1px solid ${theme.palette.divider}`,
    },
    paper: {
      width: drawerWidth,
    },
  }),
);

export function DrawerDisplay({ show, content, hideDrawer }: IDrawerDisplayProps) {
  const { drawerHeader, paper } = useStyles();
  const contentDisplay =
    content === 'contacts' ? <MyContacts /> : content === 'conversations' ? <MyConversations /> : null;
  return (
    <Drawer variant="persistent" anchor="left" open={show} onClose={hideDrawer} classes={{ paper }}>
      <Box className={drawerHeader}>
        <IconButton aria-label="collapse" onClick={hideDrawer}>
          <ArrowBackIos />
        </IconButton>
      </Box>
      <Box>{contentDisplay}</Box>
    </Drawer>
  );
}

const mapStateToProps = ({ layout }: IAppState) => ({
  show: layout.showDrawer,
  content: layout.drawerContent,
});

const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  hideDrawer: () => dispatch(hideDrawer()),
});

export const AppDrawer = connect(mapStateToProps, mapDispatchToProps)(DrawerDisplay);
