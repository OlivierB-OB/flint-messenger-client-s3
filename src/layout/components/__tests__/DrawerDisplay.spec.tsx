import React from 'react';
import { shallow } from 'enzyme';
import { DrawerDisplay } from '../AppDrawer';
import IconButton from '@material-ui/core/IconButton';

describe('DrawerDisplay', () => {
  it('drawer shown - no content', async () => {
    const props: any = {
      show: true,
      hideDrawer: 'hideDrawer',
    };
    const component = shallow(<DrawerDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('drawer shown - contacts', async () => {
    const props: any = {
      show: true,
      content: 'contacts',
      hideDrawer: 'hideDrawer',
    };
    const component = shallow(<DrawerDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('drawer shown - conversations', async () => {
    const props: any = {
      show: true,
      content: 'conversations',
      hideDrawer: 'hideDrawer',
    };
    const component = shallow(<DrawerDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('drawer not shown', async () => {
    const props: any = {
      show: false,
      hideDrawer: 'hideDrawer',
    };
    const component = shallow(<DrawerDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should allow closing the drawer', async () => {
    const props: any = {
      show: false,
      hideDrawer: jest.fn(),
    };
    const component = shallow(<DrawerDisplay {...props} />);
    component.find(IconButton).simulate('click');
    expect(props.hideDrawer).toHaveBeenCalled();
  });
});
