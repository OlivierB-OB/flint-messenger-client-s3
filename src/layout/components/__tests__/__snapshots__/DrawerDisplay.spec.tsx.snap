// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerDisplay drawer not shown 1`] = `
<WithStyles(ForwardRef(Drawer))
  anchor="left"
  classes={
    Object {
      "paper": "makeStyles-paper-2",
    }
  }
  onClose="hideDrawer"
  open={false}
  variant="persistent"
>
  <Styled(MuiBox)
    className="makeStyles-drawerHeader-1"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-label="collapse"
      onClick="hideDrawer"
    >
      <ArrowBackIosIcon />
    </WithStyles(ForwardRef(IconButton))>
  </Styled(MuiBox)>
  <Styled(MuiBox) />
</WithStyles(ForwardRef(Drawer))>
`;

exports[`DrawerDisplay drawer shown - contacts 1`] = `
<WithStyles(ForwardRef(Drawer))
  anchor="left"
  classes={
    Object {
      "paper": "makeStyles-paper-2",
    }
  }
  onClose="hideDrawer"
  open={true}
  variant="persistent"
>
  <Styled(MuiBox)
    className="makeStyles-drawerHeader-1"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-label="collapse"
      onClick="hideDrawer"
    >
      <ArrowBackIosIcon />
    </WithStyles(ForwardRef(IconButton))>
  </Styled(MuiBox)>
  <Styled(MuiBox)>
    <Connect(ContactList) />
  </Styled(MuiBox)>
</WithStyles(ForwardRef(Drawer))>
`;

exports[`DrawerDisplay drawer shown - conversations 1`] = `
<WithStyles(ForwardRef(Drawer))
  anchor="left"
  classes={
    Object {
      "paper": "makeStyles-paper-2",
    }
  }
  onClose="hideDrawer"
  open={true}
  variant="persistent"
>
  <Styled(MuiBox)
    className="makeStyles-drawerHeader-1"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-label="collapse"
      onClick="hideDrawer"
    >
      <ArrowBackIosIcon />
    </WithStyles(ForwardRef(IconButton))>
  </Styled(MuiBox)>
  <Styled(MuiBox)>
    <Connect(ConversationList) />
  </Styled(MuiBox)>
</WithStyles(ForwardRef(Drawer))>
`;

exports[`DrawerDisplay drawer shown - no content 1`] = `
<WithStyles(ForwardRef(Drawer))
  anchor="left"
  classes={
    Object {
      "paper": "makeStyles-paper-2",
    }
  }
  onClose="hideDrawer"
  open={true}
  variant="persistent"
>
  <Styled(MuiBox)
    className="makeStyles-drawerHeader-1"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-label="collapse"
      onClick="hideDrawer"
    >
      <ArrowBackIosIcon />
    </WithStyles(ForwardRef(IconButton))>
  </Styled(MuiBox)>
  <Styled(MuiBox) />
</WithStyles(ForwardRef(Drawer))>
`;
